### DISCUSSION TITLE
jsPsych.randomization.sampleWithoutReplacement - saving output?

### DISCUSSION THREAD

**Discussion by chester108** [jspsych/jsPsych#1463] [Q&A] [answered]
I have the following code:
```javascript
var feedback= [];
var feedback_one = {
    type: "image-keyboard-response",
    trial_duration: 2000,
    stimulus:function(){
        var feedback1; 
        var last_trial_correct=jsPsych.data.get().last(1).values()[0].correct;
        if (last_trial_correct){
            return feedback1 = jsPsych.randomization.sampleWithoutReplacement(button_pressed, 1); //Return 1 stimuli sampled from the randomized button pressed array without replacement
        } else {
            return feedback1 = jsPsych.randomization.sampleWithoutReplacement(button_notpressed, 1) //Return 1 stimuli sampled from the randomized button not pressed array without replacement
    }
    data.correct=feedback1;
    feedback.push(feedback1);
```

However, in my code I receive an error that states "data.correct=feedback1" is unreachable. I want to return an array of the randomly presented stimuli.

**Comment by becky-gilbert** [comment, marked as answer]
Hi @chester108, I'm not sure based on the code you provided, but I think the last two lines are supposed to be inside of the stimulus function - is that right? If so, then I don't think those lines will be called because you're returning from the function before that point in the code. 
Also, you don't have access to the `data` object from inside the `stimulus` function, so you can't update the data in a `stimulus` function in the way that you can from inside a trial's `on_finish` function.
I'm not sure if this would work, but you could try moving the code from your `stimulus` function into the [trial's `on_start` parameter](https://www.jspsych.org/overview/trial/#the-on_start-event). This function will run before the trial starts and has access to the trial parameters, which allows you to modify the trial data and the `stimulus` value, and push the feedback onto your `feedback` array. You don't need to return anything from this function.
```javascript
var feedback= [];
var feedback_one = {
    type: "image-keyboard-response",
    trial_duration: 2000,
    on_start: function(trial){
        var feedback1; 
        var last_trial_correct=jsPsych.data.get().last(1).values()[0].correct;
        if (last_trial_correct){
            feedback1 = jsPsych.randomization.sampleWithoutReplacement(button_pressed, 1); //Return 1 stimuli sampled from the randomized button pressed array without replacement
        } else {
            feedback1 = jsPsych.randomization.sampleWithoutReplacement(button_notpressed, 1) //Return 1 stimuli sampled from the randomized button not pressed array without replacement
        }
        trial.data.correct=feedback1; // store feedback1 as 'correct' in the trial data
        trial.stimulus = feedback1;  // make feedback1 the stimulus for this trial
        feedback.push(feedback1);  // add feedback1 to the feedback array
    },
    // other trial parameters
};
```
I hope this helps! Let us know if you need further guidance.

**Reply by becky-gilbert** [reply]
One more thing: the [`randomization.sampleWithoutReplacement` function](https://www.jspsych.org/core_library/jspsych-randomization/#jspsychrandomizationsamplewithoutreplacement) always returns an _array_, even when you're only sampling one element. So you probably want to get the first (only) element from that array, and store that as `feedback1`:
```javascript
feedback1 = jsPsych.randomization.sampleWithoutReplacement(button_pressed, 1)[0];
//...
feedback1 = jsPsych.randomization.sampleWithoutReplacement(button_notpressed, 1)[0];
```

**Reply by chester108** [reply]
Hi @becky-gilbert thanks so much for your reply and for looking at this! Unfortunately, that does not seem to be working. Previously, I was able to return a stimuli at random based on the answer of the previous trial. Now, the screen is blank on the feedback trial. I've included the entire code below (hopefully this makes it more clear, apologies for not doing so sooner). I don't have any error messages in my code.
```js
var button_press = [];	
var trial_one = {
    type: 'image-button-response',
    stimulus: light_off,
    choices: ['PRESS'],
    trial_duration: 3000,
    on_finish: function(data){
    var press1;
        if(data.button_pressed== "0"){
        press1= true;
    }
       else {
        press1=false;
    }
    data.correct=press1;
    button_press.push(press1);
}
}
var feedback=[];
var feedback_one = {
    type: "image-keyboard-response",
    trial_duration: 2000,
    on_start:function(trial){
        var feedback1;
        var last_trial_correct=jsPsych.data.get().last(1).values()[0].correct;
        if (last_trial_correct){
            feedback1=jsPsych.randomization.sampleWithReplacement(button_pressed, 1,[1,1])[0];
        } else {
            feedback1=jsPsych.randomization.sampleWithReplacement(button_notpressed, 1,[1,1])[0];
    }
    trial.data.correct=feedback1;
    trial.stimulus=feedback1;
    feedback.push(feedback1);
    }
}
timeline.push(trial_one,feedback_one);

```

**Reply by chester108** [reply]
@becky-gilbert - when I try "on_load" instead of "on_start", the trial progresses albeit it still does not show the feedback stimulus (screen remains blank for the duration of the trial 2000ms). It appears the issue may be in here? As you mentioned "stimulus" was not appropriate for this setting but I'm not sure what else to try.

**Reply by becky-gilbert** [reply]
Hi @chester108, thanks for posting your code! 

One problem is that there's an error in the `sampleWithReplacement` function. That function only takes two arguments: the array to sample from, and the number of samples. So you should remove the `,[1,1]` from inside the parentheses.

Another problem is that you still have to give initial values for any parameters that you want to change in the `on_start` function (sorry, I didn't know that!). So in your case, the `feedback1` trial needs to have initial values for `stimulus` and `data`. This version works for me (with some additional changes to make it reproducible).
```js
var timeline = [];

var button_pressed = ['button pressed stim 1', 'button pressed stim 2'];
var button_notpressed = ['button not pressed stim 1', 'button not pressed stim 2'];
var feedback=[];
var button_press = [];	

var trial_one = {
  type: 'image-button-response',
  stimulus: 'jsPsych-6.1.0/examples/img/1.gif',
  choices: ['PRESS'],
  trial_duration: 3000,
  on_finish: function(data){
    var press1;
    if (data.button_pressed == "0") {
      press1 = true;
    } else {
      press1 = false;
    }
    data.correct = press1;
    button_press.push(press1);
  }
}

var feedback_one = {
  type: "html-keyboard-response",
  trial_duration: 2000,
  stimulus: '',
  data: {},
  on_start:function(trial){
    var feedback1;
    var last_trial_correct=jsPsych.data.get().last(1).values()[0].correct;
    if (last_trial_correct) {
        feedback1=jsPsych.randomization.sampleWithReplacement(button_pressed, 1)[0];
    } else {
        feedback1=jsPsych.randomization.sampleWithReplacement(button_notpressed, 1)[0];
    }
    trial.stimulus = feedback1;
    trial.data.correct = feedback1;
    feedback.push(feedback1);
  }
}
timeline.push(trial_one,feedback_one);

/* start the experiment */
jsPsych.init({
  timeline: timeline,
  preload_images: ['jspsych-6.1.0/examples/img/1.gif'],
  on_finish: function() {
    console.log('button press array: ', button_press);
    console.log('feedback array: ', feedback);
    jsPsych.data.displayData();
  }
});
```

**Reply by chester108** [reply]
Thanks- @becky-gilbert that is really helpful and you've helped me learn so much!! I really appreciate it. 

If the feedback is not dependent upon the previous trials' answer and the stimuli are binary, the following code (adapted from your previous response) can also be used. I am adding this code below for the communities' reference if people run into a similar question--
```
var feedback=[];
var feedback_one = {
    type: "image-keyboard-response",
    stimulus: jsPsych.randomization.sampleWithReplacement(button_feedback, 1)[0],
    trial_duration: 2000,
    on_finish: function(data){
    var feedback1;
        if(feedback_one.stimulus== light_on){
        feedback1= true;
    }
       else {
        feedback1=false;
    }
    data.correct=feedback1;
    feedback.push(feedback1);
}
}
    timeline.push(feedback_one);
```
